keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,519 @@
+This would allow smarter routing based on the k8s object KAS is processing
+or which user/tenant tries to initiate the request, etc.
+
+### Proxy gRPC definition
+
+In order to serve a proxy request, one gRPC bidirectional stream on proxy
+server is created to serve it. It's a 1:1 mapping from TCP connection to a
+gRPC stream, so the state of TCP connection is exactly the same as the gRPC
+stream state.
+
+```grpc
+syntax = "proto3";
+
+service ProxyService {
+  // Proxy a TCP connection to a remote address defined by ConnectParam.
+  // The ConnectParam is defined in metadata under key "x-kube-net-proxy".
+  // metadata["x-kube-net-proxy"] = base64.Encode(proto.Marshal(connectOptions))
+  rpc Proxy(stream Payload) returns (stream Payload) {}
+}
+
+// ConnectOptions defines the remote TCP endpoint to connect
+message ConnectOptions {
+  string remote_addr = 1; // remote address to connect to. e.g. 8.8.8.8:53
+}
+
+// Payload defines a TCP payload.
+message Payload {
+  bytes data = 1;
+}
+```
+
+### Connectivity Proxy
+
+The connectivity proxy(s) can run in the same container as the KAS.
+It should run on the same machine and must run in the same flat network as the KAS.
+It listens on a port for gRPC connections from the KAS.
+This port would be for forwarding traffic to the appropriate cluster.
+It should have an administrative port speaking https.
+The administrative port serves the liveness probe and metrics.
+The liveness probe prevents a partially broken cluster
+where the KAS cannot connect to the cluster. This port also serves
+pprof debug commands and monitoring data for the proxy.
+
+### Direct Connection
+
+This connection type uses the default dialer.
+This allows use of the connectivity service without the connectivity proxy.
+This is a quick way to run the system in a “legacy” or fallback mode.
+Simple clusters (not needing network segregation) run this way to avoid the overhead
+(in latency or configuration) of the connectivity proxy.
+
+### Kubernetes API Server Outbound Requests
+
+The majority of the KAS communication originates from incoming requests.
+Here we cover the outgoing requests. This is our understanding of those requests
+and some details as to how they fit in this model. For the alpha release we
+support 'master', 'etcd' and 'cluster' connectivity service names.
+
+- **ETCD** It is possible to make etcd talk via the proxy.
+The etcd client takes a transport.
+(https://github.com/etcd-io/etcd/blob/master/client/client.go#L101)
+We will add configuration as to which proxy an etcd client should use.
+(https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go)
+This will add an extra process hop to our main scaling axis.
+We will scale test the impact and publish the results. As a precaution
+we will add an extra network configuration 'etcd' separate from ‘master’.
+Etcd requests can be configured separately from the rest of 'master'.
```

Comment:
What is the use case for that?

Reply:
I see two use cases. 1) By dictating what is supposed to be etcd destined traffic you can configure that proxy as the only one with access to etcd. This can be another layer of protection for etcd from things like our proxy sub resources. 2) Etcd is a primary dimension of scalability. This allows us to do things like bypass proxy specifically for etcd traffic if you are running close to the edge.